Check if the iterator is valid before dereferencing it

nkaretnikov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72405

Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

Differential Revision: [D34423660](https://our.internmc.facebook.com/intern/diff/D34423660)

nkaretnikov committed:
Check if the iterator is valid before dereferencing it

Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

lezcano reviewed (commented):
Great! Would you mind adding a test for this as well?

nkaretnikov committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented:
done, pls take another look

lezcano reviewed (approved):
LGTM. Can you take a look @anjali411 ?

anjali411 commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -405,7 +405,11 @@ void py_set_default_dtype(PyObject* obj) {
+    if (it != tensor_types.end()) {
+      set_default_tensor_type(*it);
+    } else {
+      throw TypeError("tensor type not found");
```
we should have the same error message as below `"invalid dtype object"`

anjali411 replied:
maybe consolidate the logic to throw error for invalid dtype:

```
bool not_valid_dtype = true;
if (THPDtype_Check(obj)) {
    ...
    if (it != tensor_types.end()) {
       not_valid_dtype = false;
      set_default_tensor_type(*it);
    }
}
if (not_valid_dtype) throw TypeError("tensor type not found");
```

anjali411 commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -405,7 +405,11 @@ void py_set_default_dtype(PyObject* obj) {
+    if (it != tensor_types.end()) {
+      set_default_tensor_type(*it);
+    } else {
+      throw TypeError("tensor type not found");
```
we should throw the error with the same message as for other non-floating point types
```
>>> torch.set_default_dtype(torch.int16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/fsx/users/chourdiaanjali/tree2/torch/__init__.py", line 369, in set_default_dtype
    _C._set_default_dtype(d)
TypeError: only floating-point types are supported as the default type
```

anjali411 replied:
I think the real issue is that the `scalar_types` in this function
https://github.com/pytorch/pytorch/blob/224093db113fd333a7cb8e5e07308de5098e4a62/torch/csrc/utils/tensor_types.cpp#L84 doesn't include complex types (and it should). Currently any dtype not in that list also leads to a segfault, for example:
```
>>> torch.set_default_dtype(torch.qint32)
Segmentation fault (core dumped)
```
. We should consider keeping this list in sync with `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS` https://github.com/pytorch/pytorch/blob/224093db113fd333a7cb8e5e07308de5098e4a62/c10/core/ScalarType.h#L29

mruberry commented:
We should also update the documentation for `set_default_dtype` to clarify what's supported

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -2121,6 +2121,11 @@ def test_constructor_dtypes(self, device):
         self.assertRaises(TypeError, lambda: torch.set_default_dtype(torch.int64))
 
+        # don't support complex.
+        self.assertRaises(TypeError, lambda: torch.set_default_dtype(torch.complex32))
```
I think we're about to remove `complex32`, anyway, so we don't need to worry about testing it

anjali411 replied:
we are only removing it in the release so we should still add it

nkaretnikov commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -405,7 +405,11 @@ void py_set_default_dtype(PyObject* obj) {
+    if (it != tensor_types.end()) {
+      set_default_tensor_type(*it);
+    } else {
+      throw TypeError("tensor type not found");
```
yep, this is a good catch, i'll see if i can refactor it somehow to avoid this problem entirely

nkaretnikov replied:
@anjali411 do we also want to support more backends in `all_declared_types`? the list of `Backend`s is larger than the one defined in the `backends` var

nkaretnikov committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented:
made the changes, please take a look. tests seem to fail for me locally tho, let's see what ci says

nkaretnikov commented:
i mentioned it in the reply to a review comment, but let me ask again just in case: do we need to have a similar macro for backends too? not all values that are part of the backend type are listed rn, or is it expected?

anjali411 commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -83,11 +83,15 @@ at::TensorOptions options_from_string(const std::string& str) {
-  std::vector<ScalarType> scalar_types = { ScalarType::Byte, ScalarType::Char, ScalarType::Double, ScalarType::Float,
-                                           ScalarType::Int, ScalarType::Long, ScalarType::Short, ScalarType::Half,
-                                           ScalarType::Bool, ScalarType::BFloat16};
+  // keep in sync with ScalarType
```
thank you!

anjali411 commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -405,7 +405,11 @@ void py_set_default_dtype(PyObject* obj) {
         return x->get_backend() == backend && x->get_scalar_type() == scalar_type;
       });
-    set_default_tensor_type(*it);
+    if (it != tensor_types.end()) {
```
I think this is unnecessary because `THPDtype_Check` takes care of it

nkaretnikov replied:
wait, but that's exactly the reason it used to segfault before. end was dereferenced, so a garbage value was passed to set_default_tensor_type, which was dereferenced again causing a segfault. am i missing something?

nkaretnikov replied:
i mean, in general you're supposed to check return values of things, so i'm in favor of keeping this regardless

anjali411 replied:
yup so I'd just add `TORCH_INTERNAL_ASSERT(it != tensor_types.end())` because we expect to find a non null value for it, and if that's not the case then there's something wrong in the internal function behavior, and not a user code issue.

nkaretnikov replied:
yep, will do

nkaretnikov replied:
done, please take a look

anjali411 commented on test/test_tensor_creation_ops.py:
```diff
@@ -2117,13 +2117,39 @@ def test_constructor_dtypes(self, device):
 
+        # don't allow passing dtype to set_default_dtype
+        # see AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS
+        all_scalar_types_with_complex_and_qints = (
```
should we add a method in `torch/testing/_legacy.py` to get the list of all valid dtypes (including quantized types)?

cc. @mruberry

nkaretnikov replied:
fwiw, i'm aware of that file. decided to list them explicitly here as it makes things easier to follow. but maybe it would be better to reuse whatever is in legacy. that said, could this set of types be used anywhere else? i'm not sure. but there are a bunch of places where people define lists like this, so dunno. lmk what you prefer

anjali411 replied:
I totally understand. I am suggester to have one getter to have one source of truth. Using these explicitly defined lists got us into this trouble in the firsts place

nkaretnikov replied:
agreed, i'll move these to legacy

nkaretnikov replied:
did the minimal required change. i think this is good enough for now. if others want to use this, they can extend code in the legacy file to match other dtypes

nkaretnikov commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -83,11 +83,15 @@ at::TensorOptions options_from_string(const std::string& str) {
 
 std::vector<std::pair<Backend, ScalarType>> all_declared_types() {
   std::vector<std::pair<Backend, ScalarType>> ret;
+
```
i did it on purpose to improve readability

nkaretnikov commented:
as for updating the docs, i'm simply not sure what is supported atm, see:
https://pytorch.org/docs/stable/generated/torch.set_default_dtype.html?highlight=set_default_dtype

> Sets the default floating point dtype to d. Supports torch.float32 and torch.float64 as inputs. Other dtypes may be accepted without complaint but are not supported and are unlikely to work as expected.

you can see from my test that these don't throw:
```
(torch.half, torch.float, torch.double, torch.bfloat16)
```

so are these two extra types actually supported now or just don't raise an error (as the doc says)?

nkaretnikov committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

anjali411 reviewed (approved):
Looks great! Thanks @nkaretnikov

nkaretnikov commented:
i see the failed checks, not quite sure why yet. will take a look and push and update.

nkaretnikov committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented:
rebased on the current upstream main branch, let's see if it fixes the errors

nkaretnikov commented:
rebased on the current upstream main branch, let's see if it fixes the errors

anjali411 commented:
@nkaretnikov the test failures on this PR seem unrelated. could you rebase it to verify if that's true?

nkaretnikov commented:
@malfet @anjali411 i'm going to rebase and see if it fixes the issues

Commit:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented:
@malfet @anjali411 rebased, waiting for ci

nkaretnikov commented:
@malfet @anjali411 hmm, looks like there are still issues in the docs/pull stages. do you know why? i tried looking at this a few days ago and i couldn't figure it out. feel free to take over if you know what to do, or cc relevant ppl. if you fix this, please tell me how, so i could fix similar problems in the future. thx

anjali411 committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

anjali411 committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

anjali411 committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

anjali411 committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

anjali411 committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

anjali411 committed:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

Commit:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented on test/test_tensor_creation_ops.py:
```diff
@@ -2117,13 +2117,24 @@ def test_constructor_dtypes(self, device):
-        self.assertRaises(TypeError, lambda: torch.set_default_dtype(torch.int64))
-
         # don't allow passing dtype to set_default_tensor_type
+        self.assertRaises(TypeError, lambda: torch.set_default_tensor_type('torch.IntTensor'))
```
added a new comment instead of removing and added closer to other relevant code

nkaretnikov commented:
@anjali411 made another round of changes, pls take a look once ci is done, ty!

nkaretnikov commented:
there's already one failing test, will look at it on sunday probably

nkaretnikov commented:
didn't have time on the weekend at all, starting to look at it now

peterbell10 commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -83,11 +83,18 @@ at::TensorOptions options_from_string(const std::string& str) {
+  std::vector<ScalarType> scalar_types = {
+    #define DEFINE_ENUM(_1, n) ScalarType::n,
+      AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(DEFINE_ENUM)
+  };
```
I think `torch.ComplexDoubleTensor`, `torch.cuda.ComplexDoubleTensor` etc. intentionally wasn't added because those types are legacy. Instead, I think `py_set_default_dtype` should avoid the intermediate step through the legacy tensor type api.

nkaretnikov replied:
ouch, i pushed a new set of changes before seeing this comment

@anjali411 wdyt? do we rollback to the previous set of types and just add more asserts in the function that failed initially? either way, the thing that we're doing rn is error-prone because it's not actually all-types (the macros are different). so if more types are added later, it'll fall apart again :/

peterbell10 replied:
@anjali411 just to be 100% clear. This PR has the non-obvious effect of adding 4 new classes to the public python API: `torch.ComplexFloatTensor`, `torch.ComplexDoubleTensor`, `torch.cuda.ComplexFloatTensor` and `torch.cuda.ComplexDoubleTensor`.

I'm not sure this was the desired outcome, so want to make sure.

Lezcano replied:
cc @albanD

Commit:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented:
i rebased and added a skip for the doc generation. not sure if it fixes all the problems (i cannot run this test locally because i cannot access s3, for instance: https://github.com/pytorch/pytorch/runs/5252447167). also see peter's comment above

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -2099,6 +2099,9 @@ def test_constructor_dtypes(self, device):
         self.assertIs(torch.FloatStorage, torch.Storage)
 
+        # only floating-point types are supported as the default type
+        self.assertRaises(TypeError, lambda: torch.set_default_tensor_type('torch.IntTensor'))
```
Isn't this redundant with the below tests?

mruberry replied:
Oh - this is `set_default_tensor_type`, not `set_default_dtype` -- my mistake

Commit:
Update on "Check if the iterator is valid before dereferencing it"


Fixes #71674.

This shouldn't segfault now:

```
import torch
d = torch.complex64
torch.set_default_dtype(d)
```

[ghstack-poisoned]

nkaretnikov commented:
the windows failures seemed unrelated, rebased again to confirm

anjali411 reviewed (approved):
LGTM thanks @nkaretnikov

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34423660).

nkaretnikov commented:
oops, totally forgot about the labels, sorry!

nkaretnikov commented:
@anjali411 what would be the proper release label for this one? i'm struggling to find the matching one in the list

nkaretnikov commented:
i've used the complex label since this seems to be the most specific one related to exposed changes here. feel free to change if you disagree
